### PR TITLE
Tf timeout

### PR DESCRIPTION
**Pull request by Marwan99**
## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #978 |
| Primary OS tested on | (Ubuntu 18.04) |

---

## Description of contribution in a few bullet points
* Added the required tf timeouts
* Refactored tf timeouts to use `transform_tolerance` parameter
* Refactored  amcl former `transform_tolerance` parameter to be `transform_publish_shift` (this is a proposal, just to differentiate this parameter from `transform_tolerance`)

## Description of documentation updates required from your changes

* Update amcl param list with  `transform_publish_shift` and  `transform_tolerance` (if proposal accepted)

---

## Future work 

* Refactor map->base_link or map->odom transform to use `getCurrentPose`

### CODE REVIEW

File: nav2_costmap_2d/plugins/static_layer.cpp [hunk 1]
```diff
@@ -133,10 +134,13 @@ StaticLayer::getParameters()
   node_->get_parameter("lethal_cost_threshold", temp_lethal_threshold);
   node_->get_parameter("unknown_cost_value", unknown_cost_value_);
   node_->get_parameter("trinary_costmap", trinary_costmap_);
+  node_->get_parameter("transform_tolerance", temp_tf_tol);
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/static_layer.cpp:137]: I think this has to be declared.

File: nav2_amcl/src/amcl_node.cpp [hunk 2]
```diff
@@ -1088,6 +1093,7 @@ AmclNode::initParameters()
   get_parameter("sigma_hit", sigma_hit_);
   get_parameter("tf_broadcast", tf_broadcast_);
   get_parameter("transform_tolerance", tmp_tol);
+  get_parameter("transform_publish_shift", tmp_tf_shift);
```
Comment (SteveMacenski) [inline nav2_amcl/src/amcl_node.cpp:1096]: Do you think both can use the same parameter? I think within the scope of AMCL that's OK and use `transform_tolerance`

File: nav2_costmap_2d/plugins/static_layer.cpp (hunk 1 above)
Comment (Marwan99) [reply nav2_costmap_2d/plugins/static_layer.cpp:137]: Sure, will fix it.

File: nav2_amcl/src/amcl_node.cpp (hunk 2 above)
Comment (Marwan99) [reply nav2_amcl/src/amcl_node.cpp:1096]: Will change it back, np.

**Comment by SteveMacenski** [comment]
Also make sure to run the linters. If you run tests locally it'll trigger those too, but you can trigger just specific linters to iterate with `ament_cpplint` `ament_uncrustify` and `ament_flake8` in the terminal

File: nav2_costmap_2d/plugins/static_layer.cpp [hunk 3]
```diff
@@ -117,10 +117,12 @@ void
 StaticLayer::getParameters()
 {
   int temp_lethal_threshold = 0;
+  double temp_tf_tol;
 
   declareParameter("enabled", rclcpp::ParameterValue(true));
   declareParameter("subscribe_to_updates", rclcpp::ParameterValue(false));
   declareParameter("map_subscribe_transient_local", rclcpp::ParameterValue(true));
+  declareParameter("transform_tolerance", rclcpp::ParameterValue(1.0));
```
Comment (SteveMacenski) [inline nav2_costmap_2d/plugins/static_layer.cpp:125]:
```suggestion
  declareParameter("transform_tolerance", rclcpp::ParameterValue(0.0));
```

Comment (SteveMacenski) [reply nav2_costmap_2d/plugins/static_layer.cpp:125]: Just to keep default behavior the same.

File: nav2_amcl/include/nav2_amcl/amcl_node.hpp [hunk 4]
```diff
@@ -243,7 +243,8 @@ class AmclNode : public nav2_util::LifecycleNode
   tf2::Duration save_pose_period_;
   double sigma_hit_;
   bool tf_broadcast_;
-  tf2::Duration transform_tolerance_;
+  tf2::Duration transform_tolerance_; // tf timeout duration
+  tf2::Duration transform_publish_shift_; // post-date the transform that is published, this was formerly called transform_tolerance_
```
Comment (SteveMacenski) [inline nav2_amcl/include/nav2_amcl/amcl_node.hpp:247]: Please remove this and just use transform_tolerance_ for all in AMCL

**Comment by SteveMacenski** [comment]
@Marwan99 close - just 2 more things and passing CI and we can move on!

File: nav2_amcl/src/amcl_node.cpp (hunk 2 above)
Comment (Marwan99) [reply nav2_amcl/src/amcl_node.cpp:1096]: If we are using the same name, do you think we need to update the parameter description?

> Time with which to post-date the transform that is published, to indicate that this transform 
> is valid into the future

File: nav2_amcl/include/nav2_amcl/amcl_node.hpp (hunk 4 above)
Comment (Marwan99) [reply nav2_amcl/include/nav2_amcl/amcl_node.hpp:247]: Coming in the next commit

**Comment by SteveMacenski** [comment]
I think this is missing the observation buffer timeouts

**Comment by SteveMacenski** [comment]
What are your thoughts on `getCurrentPose` timeouts? Leave as is, modify them to use?

**Comment by Marwan99** [comment]
Probably modify them as well to provide flexibility across the board.

**Comment by Marwan99** [comment]
Will run linters to pass CI.

**Comment by SteveMacenski** [comment]
OK, should that be done in this PR or a follow up one?

**Comment by Marwan99** [comment]
I guess in this PR should be fine, still within scope.

**Comment by SteveMacenski** [comment]
@Marwan99 lets do a second PR. 2 PRs are better than 1 large PR in my book.

If you look at the CI tests, you still have some linting errors before we can merge this

**Comment by SteveMacenski** [comment]
@Marwan99 ping me when this passes tests and I'll merge

**Pull request merged**